test/jdk/java/lang/Thread/virtual/MonitorEnterWaitOOME.java

```diff
@@ -0,0 +1,139 @@
+ * @library /test/lib
+ * @run main/othervm -XX:+UseG1GC -Xmx48M MonitorEnterWaitOOME true 0
+ */
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class MonitorEnterWaitOOME {
+    static volatile Object data;
+    static Thread.State dummyState = Thread.State.RUNNABLE; // load java.lang.Thread$State
+
+    public static void main(String[] args) throws Throwable {
+        final boolean testWait = args.length >= 1 ? Boolean.parseBoolean(args[0]) : false;
+        final long timeout = testWait && args.length == 2 ? Long.parseLong(args[1]) : 0L;
+
+        Thread vthread;
+        var lock = new Object();
+        var canFillHeap = new AtomicBoolean();
+        var heapFilled = new AtomicBoolean();
+        var heapCollected = new AtomicBoolean();
+        var exRef = new AtomicReference<Throwable>();
+        synchronized (lock) {
+            vthread = Thread.ofVirtual().start(() -> {
+                try {
+                    awaitTrue(canFillHeap);
+                    data = fillHeap();
+                    heapFilled.set(true);
+                    synchronized (lock) {
+                        if (testWait) {
+                            lock.wait(timeout);
+                        }
+                    }
+                    data = null;
+                    System.gc();
+                    heapCollected.set(true);
+                } catch (Throwable e) {
+                    data = null;
+                    System.gc(); // avoid nested OOME
+                    exRef.set(e);
+                }
+            });
+            canFillHeap.set(true);
+            awaitTrue(heapFilled);
+            awaitState(vthread, Thread.State.BLOCKED);
+        }
+        if (testWait && timeout == 0) {
+            awaitState(vthread, Thread.State.WAITING);
+            synchronized (lock) {
+                lock.notify();
+            }
+        }
+        joinVThread(vthread, heapCollected, exRef);
+        assert exRef.get() == null;
+    }
+
+    private static Object[] fillHeap() {
+        Object[] first = null, last = null;
+        int size = 1 << 20;
+        while (size > 0) {
+            try {
+                Object[] array = new Object[size];
+                if (first == null) {
+                    first = array;
+                } else {
+                    last[0] = array;
+                }
+                last = array;
+            } catch (OutOfMemoryError oome) {
+                size = size >>> 1;
+            }
+        }
+        return first;
+    }
+
+    private static void awaitTrue(AtomicBoolean ready) {
+        // Don't call anything that might allocate from the Java heap.
+        while (!ready.get()) {}
+    }
+
+    private static void awaitState(Thread thread, Thread.State expectedState) {
+        // Don't call anything that might allocate from the Java heap.
+        while (thread.getState() != expectedState) {}
```

Comment:
I think I would put Thread.onSpinWait() in this loop to reduce the overhead of the spin.

Reply:
I added it here and in the other loops.